File: _services/algorithm-lab/tests/diagnostics.py
"""Check actual worker diagnostics against known student source coordinates."""
import json
from pathlib import Path
root=Path(__file__).resolve().parents[3]
worker=(root/'algorithm-lab/worker.js').read_text(encoding='utf-8')
grader=worker.split('const answer = py.runPython(`',1)[1].split('`);',1)[0]
def run(code):
 scope={'_payload_json':json.dumps({'code':code,'problem':{'function':'solve'},'cases':[{'args':[1],'expected':1,'public':True}]})}
 exec(compile(grader,'worker.py','exec'),scope)
 return scope['_answer']
r=run('def solve(x):\n    return x +\n')
assert r['diagnostics'][0]['line']==2 and r['diagnostics'][0]['column']
r=run('def helper(x):\n    return 1 / 0\n\ndef solve(x):\n    return helper(x)\n')
d=r['rows'][0]['diagnostic']
assert d['line']==2 and [f['line'] for f in d['frames']]==[5,2]
assert 'ZeroDivisionError' in d['raw'] and d['file']=='answer.py'
r=run('def solve(x):\n    raise NotImplementedError\n')
assert r['rows'][0]['diagnostic']['line']==2
r=run('def other(x):\n    return x\n')
assert r['diagnostics'][0]['line'] is None
r=run('def solve(x):\n    return x\n')
assert r['passed']==1 and r['rows'][0]['diagnostic'] is None
print('Python syntax, runtime, nested calls, unfinished answer, missing function, and success diagnostics passed.')

File: _services/algorithm-lab/tests/diagnostics.mjs
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
import vm from 'node:vm';
const lab=path.resolve(import.meta.dirname,'../../../algorithm-lab');
let receive;
const next=()=>new Promise(resolve=>{receive=resolve;});
const ctx=vm.createContext({TextEncoder,TextDecoder,WebAssembly,Uint8Array,ArrayBuffer,DataView,console,setTimeout,clearTimeout,performance,
 fetch:async file=>{const b=fs.readFileSync(path.join(lab,String(file).split('?')[0]));return {ok:true,arrayBuffer:async()=>b.buffer.slice(b.byteOffset,b.byteOffset+b.byteLength)};},
 self:{postMessage:data=>{if(data.type!=='phase')receive(data);}}
});
ctx.importScripts=(...files)=>files.forEach(file=>vm.runInContext(fs.readFileSync(path.join(lab,file.split('?')[0]),'utf8'),ctx));
const boot=next();vm.runInContext(fs.readFileSync(path.join(lab,'c-worker.js'),'utf8'),ctx);assert.equal((await boot).type,'ready');
const p=JSON.parse(fs.readFileSync(path.join(lab,'problems.json'),'utf8'))[0];
async function run(code){const pending=next();await ctx.self.onmessage({data:{token:1,problem:p,cases:p.tests.slice(0,1),code}});return (await pending).report;}
let r=await run('int linear_search(const int A[], int n, int target) {\n  return 0\n}\n');
assert.ok(r.error);assert.equal(r.diagnostics.find(d=>d.severity==='error').line,2);assert.equal(r.diagnostics[0].file,'answer.c');assert.match(r.rawError,/expected ';'/);
r=await run('int linear_search(const int A[], int n, int target) {\n  int length = sizeof(A) / sizeof(A[0]);\n  return length;\n}\n');
assert.ok(!r.error);assert.ok(r.diagnostics.some(d=>d.severity==='warning'&&d.line===2));
r=await run('int linear_search(const int A[], int n, int target) {\n  __builtin_trap();\n  return 0;\n}\n');
assert.ok(r.error);assert.equal(r.diagnostics[0].line,null);assert.match(r.rawError,/unreachable|trap/);
r=await run('int linear_search(const int A[], int n, int target) {\n  for(int i=0;i<n;i++) if(A[i]==target) return i;\n  return -1;\n}\n');
assert.ok(!r.error);assert.equal(r.passed,1);
console.log('Real Clang/WASM: exact student compile line, warnings, runtime trap without invented line, and success passed.');
